extensions/functions_comparison.yaml

```diff
@@ -83,4 +83,89 @@ scalar_functions:
       - args:
           - value: fp64
         return: BOOLEAN
+  -
+    name: "least"
+    description: >-
+      Returns the smallest value from a list of arguments. Only return null if 'all' arguments evaluate to null.
+
+      String comparison is done in lexicographical ordering, one character at a time, from left to right.
+      Uppercase letters are less than lowercase letters.
+    impls:
+      - args:
+          - value: i8
+        variadic:
+          min: 1
+        return: i8
+      - args:
+          - value: i16
+        variadic:
+          min: 1
+        return: i16
+      - args:
+          - value: i32
+        variadic:
+          min: 1
+        return: i32
+      - args:
+          - value: i64
+        variadic:
+          min: 1
+        return: i64
+      - args:
+          - value: fp32
+        variadic:
+          min: 1
+        return: fp32
+      - args:
+        - value: fp64
+        variadic:
+          min: 1
+        return: fp64
+      - args:
+        - value: "string"
+        variadic:
+          min: 1
+        return: "string"
```

Comment:
Should we add the rest of the types here?

I would add everything except:

* `interval_year`
* `interval_day`
* `struct`
* `list`
* `map`

Reply:
Just added all the other ones, except for the ones you listed and boolean.

Reply:
@cpcloud Why don't you consider intervals to be comparable? Substrait effectively defines them as a number of months and a number of microseconds respectively, so I don't see why they're special. I would personally argue that only UUIDs and maps make little sense to order, though it's perfectly possible to define an ordering for them (for maps because they are basically just defined to behave like `list<struct<K, V>>`, and both of those can be compared using tie).

IMO this should just be `least(T) -> T`/`greatest(T) -> T`. Likewise for all normal comparison functions. If an engine doesn't consider a type orderable, they will already have had to solve this problem and return suitable errors for sort relations with unorderable sort keys and no custom comparator function.

Reply:
This kind of slipped through the cracks because I thought we had resolved this offline (more or less), but I must admit I never considered this example.

At some point in the past, when I clarified what the built-in types mean, I made the assertion that it should be allowed to store interval types using a single number. In other words, 1 day, 24 hours, 1440 minutes, or 86400 seconds, should all mean exactly the same thing. I figured we could do this because year-month and day-second were separate types already anyway (but, again, didn't consider this one). I did this to make the description of the types at all compatible with Arrow's types, because you could fairly easily construct an interval type that stores the components separately using structs anyway, and frankly, because (even after this example) I remain convinced that it's good enough.

For this particular example, I would argue that the result depends on whether you're adding the interval to a `timestamp` or `timestamp_tz`. `timestamp` has no timezone awareness, so in both cases you would get `2022-01-02 12:00:00`. `timestamp_tz` instead represents "real" time, where DST just doesn't exist. You simply get the timestamp that occurs 24 hours/1 day later. *When represented in this particular timezone* that might result in `2022-01-01 12:00:00` -> `2022-01-02 13:00:00`, but represented in UTC it might be `2022-01-01 11:00:00Z` -> `2022-01-02 11:00:00Z` if that timezone happened to be CET. Timezone shenanigans in general are captured by the conversion between `timestamp` and `timestamp_tz`, and need not exist anywhere else.

This leaves only the more fundamental problem that months are not always the same length in our calendar, which is already covered by having different types for year-month and day-second, and the lack of overlap between these ranges. Even leap seconds (if we would want to consider those) are defined to always happen at the end of a month.

What I *don't* know is whether existing query engines and SQL operations are defined with sufficient sanity to be encompassed by this logic. I'm going to hazard a guess based on recent experiences with null and say no, so maybe we need to revise these types again. But in any case, the current definition of Substrait interval types is encompassed by a number with some implicit time unit associated with it (i.e. seconds or months), which makes them trivially ordered.